### PR TITLE
Implement cluster autoscaler as bootstrap addon

### PR DESCRIPTION
**Pull request by olemarkus**
We see that when managing a larger number of clusters, getting the cluster autoscaler version correct is a bit of a hassle. We also need to take care to properly update CAS after cluster upgrades etc. So adding CAS as a configurable bootstrap addon seems like a good idea to make this more manageable.

If no one objects to this, I will finish this and add documentation and validation.

### CODE REVIEW

**Comment by MoShitrit** [comment]
That's a great idea, thanks for adding that @olemarkus !!

**Comment by MoShitrit** [comment]
BTW, I guess it would be also worth adding some documentation for it? Or will that be done through a separate PR?

**Comment by olemarkus** [comment]
Docs will come as part of this PR. Just want to ensure there are no objections before starting on that.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 1]
```diff
@@ -444,18 +444,39 @@ func (b *BootstrapChannelBuilder) buildAddons() *channelsapi.Addons {
 	if kubeDNS.Provider == "CoreDNS" && NodeLocalDNS != nil && fi.BoolValue(NodeLocalDNS.Enabled) {
 		{
 			key := "nodelocaldns.addons.k8s.io"
-			version := "1.18.0"
+			version := "1.19.0"
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:447]: This looks like an unrelated concern.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 2]
```diff
@@ -126,6 +126,7 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap, secretStore fi.SecretS
 
 	// will return openstack external ccm image location for current kubernetes version
 	dest["OpenStackCCMTag"] = tf.OpenStackCCMTag
+	dest["CasTag"] = tf.CasTag
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:129]: This seems a bit terse. Perhaps `ClusterAutoscalerTag`

File: upup/pkg/fi/cloudup/template_functions.go [hunk 3]
```diff
@@ -526,3 +527,23 @@ func (tf *TemplateFunctions) OpenStackCCMTag() string {
...
+		case 19:
+			tag = "v1.19.0"
+		case 18:
+			tag = "v1.18.2"
+		case 17:
+			tag = "v1.17.3"
+		case 16:
+			tag = "v1.16.6"
+		case 15:
+			tag = "v1.15.7"
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:545]: I think I'd prefer an error for the default case. Otherwise we might forget to update it for a new Kubernetes minor version.

File: pkg/model/components/clusterautoscaler.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	if cas.Expander == nil {
+		cas.Expander = fi.String("least-waste")
+	}
+	if cas.ScaleDownUtilizationThreshold == nil {
+		cas.ScaleDownUtilizationThreshold = fi.String("0.5")
+	}
+	if cas.SkipNodesWithLocalStorage == nil {
+		cas.SkipNodesWithLocalStorage = fi.Bool(true)
```
[the remark this answers is not in the document]
Comment (olemarkus) [reply pkg/model/components/clusterautoscaler.go:46]: 100%. It is a bad default from upstream.

File: docs/cluster_spec.md [hunk 5]
```diff
@@ -616,6 +616,20 @@ spec:
...
+## Cluster autoscaler
+
+{{ kops_feature_table(kops_added_default='1.19', k8s_min='1.15') }}
+
+Cluster autoscaler can be enabled to automatically adjust the size of the kubernetes cluster.
+
+```yaml
+spec:
+  cluster:
+    clusterAutoscaler: {}
```
Comment (johngmyers) [inline docs/cluster_spec.md:628]: Would be better to give an example that disables the two `SkipNodesWith` options.

File: pkg/apis/kops/cluster.go [hunk 6]
```diff
@@ -191,6 +191,9 @@ type ClusterSpec struct {
 	SysctlParameters []string `json:"sysctlParameters,omitempty"`
 	// RollingUpdate defines the default rolling-update settings for instance groups
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
+
+	// ClusterAutoscaler defines the cluaster autoscaler configuration
```
Comment (johngmyers) [inline pkg/apis/kops/cluster.go:195]:
```suggestion
	// ClusterAutoscaler defines the cluaster autoscaler configuration.
```

File: pkg/apis/kops/componentconfig.go [hunk 7]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
 	Openstack *OpenstackConfiguration `json:"openstack,omitempty"`
 }
 
+// ClusterAutoscalerConfig determines the cluster autoscaler configuration
+type ClusterAutoscalerConfig struct {
+	// Expander determines the strategy for which instance group gets expanded
+	// Supported values: least-waste, most-pods, random
+	// Default: least-waste
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:762]:
```suggestion
	// Expander determines the strategy for which instance group gets expanded.
	// Supported values: least-waste, most-pods, random.
	// Default: least-waste
```

Otherwise the generated docs look odd.

File: pkg/apis/kops/componentconfig.go [hunk 8]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
...
+	Expander *string `json:"expander,omitempty"`
+	// BalanceSimilarNodeGroups makes cluster autoscaler treat similar node groups as one
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:764]:
```suggestion
	// BalanceSimilarNodeGroups makes cluster autoscaler treat similar node groups as one.
```

File: pkg/apis/kops/componentconfig.go [hunk 9]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
...
+	// Default: false
+	BalanceSimilarNodeGroups *bool `json:"balanceSimilarNodeGroups,omitempty"`
+	// ScaleDownUtilizationThreshold determines the utilization threshold for node scale-down
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:767]:
```suggestion
	// ScaleDownUtilizationThreshold determines the utilization threshold for node scale-down.
```

File: pkg/apis/kops/componentconfig.go [hunk 10]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
...
+	// Default: 0.5
+	ScaleDownUtilizationThreshold *string `json:"scaleDownUtilizationThreshold,omitempty"`
+	// SkipNodesWithSystemPods makes cluster autoscaler skip scale-down of nodes with system pods
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:770]:
```suggestion
	// SkipNodesWithSystemPods makes cluster autoscaler skip scale-down of nodes with non-DaemonSet pods in the kube-system namespace.
```

File: pkg/apis/kops/componentconfig.go [hunk 11]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
...
+	// Default: true
+	SkipNodesWithSystemPods *bool `json:"skipNodesWithSystemPods,omitempty"`
+	// SkipNodesWithLocalStorage makes cluster autoscaler skip scale-down of nodes with local storage
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:773]:
```suggestion
	// SkipNodesWithLocalStorage makes cluster autoscaler skip scale-down of nodes with local storage.
```

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 12]
```diff
@@ -189,6 +189,9 @@ type ClusterSpec struct {
 	SysctlParameters []string `json:"sysctlParameters,omitempty"`
 	// RollingUpdate defines the default rolling-update settings for instance groups
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
+
+	// ClusterAutoscaler defines the cluaster autoscaler configuration
```
Comment (johngmyers) [inline pkg/apis/kops/v1alpha2/cluster.go:193]:
```suggestion
	// ClusterAutoscaler defines the cluaster autoscaler configuration.
```

File: pkg/apis/kops/componentconfig.go [hunk 13]
```diff
@@ -755,6 +755,26 @@ type CloudConfiguration struct {
 	Openstack *OpenstackConfiguration `json:"openstack,omitempty"`
 }
 
+// ClusterAutoscalerConfig determines the cluster autoscaler configuration
```
Comment (johngmyers) [inline pkg/apis/kops/componentconfig.go:758]:
```suggestion
// ClusterAutoscalerConfig determines the cluster autoscaler configuration.
```

File: pkg/apis/kops/v1alpha2/componentconfig.go [hunk 14]
```diff
@@ -756,6 +756,26 @@ type CloudConfiguration struct {
...
+	// BalanceSimilarNodeGroups makes cluster autoscaler treat similar node groups as one
+	// Default: false
+	BalanceSimilarNodeGroups *bool `json:"balanceSimilarNodeGroups,omitempty"`
+	// ScaleDownUtilizationThreshold determines the utilization threshold for node scale-down
+	// Default: 0.5
+	ScaleDownUtilizationThreshold *string `json:"scaleDownUtilizationThreshold,omitempty"`
+	// SkipNodesWithSystemPods makes cluster autoscaler skip scale-down of nodes with system pods
+	// Default: true
+	SkipNodesWithSystemPods *bool `json:"skipNodesWithSystemPods,omitempty"`
+	// SkipNodesWithLocalStorage makes cluster autoscaler skip scale-down of nodes with local storage
```
Comment (johngmyers) [inline pkg/apis/kops/v1alpha2/componentconfig.go:774]: Same comment fixups as for the non-versioned API.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 15]
```diff
@@ -447,15 +447,36 @@ func (b *BootstrapChannelBuilder) buildAddons() *channelsapi.Addons {
 			version := "1.18.0"
 
 			{
-				location := key + "/k8s-1.12.yaml"
-				id := "k8s-1.12"
+				location := key + "/k8s-1.15.yaml"
+				id := "k8s-1.15"
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:451]: This still looks like an unrelated change.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 16]
```diff
@@ -526,3 +528,34 @@ func (tf *TemplateFunctions) OpenStackCCMTag() string {
 	}
 	return tag
 }
+
+// GetInstanceGroup returns the instance group with the specified name
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:532]:
```suggestion
// GetInstanceGroups returns a map containing the defined instance groups.
```

File: k8s/crds/kops.k8s.io_clusters.yaml [hunk 17]
```diff
@@ -335,6 +335,28 @@ spec:
               cloudProvider:
                 description: The CloudProvider to use (aws or gce)
                 type: string
+              clusterAutoscaler:
+                description: ClusterAutoscaler defines the cluaster autoscaler configuration.
```
Comment (gjtempleton) [inline k8s/crds/kops.k8s.io_clusters.yaml:339]:
```suggestion
                description: ClusterAutoscaler defines the cluster autoscaler configuration.
```

File: pkg/apis/kops/cluster.go [hunk 18]
```diff
@@ -191,6 +191,9 @@ type ClusterSpec struct {
 	SysctlParameters []string `json:"sysctlParameters,omitempty"`
 	// RollingUpdate defines the default rolling-update settings for instance groups
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
+
+	// ClusterAutoscaler defines the cluaster autoscaler configuration.
```
Comment (gjtempleton) [inline pkg/apis/kops/cluster.go:195]:
```suggestion
	// ClusterAutoscaler defines the cluster autoscaler configuration.
```

File: pkg/model/components/clusterautoscaler.go [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+func (b *ClusterAutoscalerOptionsBuilder) BuildOptions(o interface{}) error {
+	clusterSpec := o.(*kops.ClusterSpec)
+	cas := clusterSpec.ClusterAutoscaler
+	if cas == nil || !fi.BoolValue(cas.Enabled) {
+		return nil
+	}
+
+	if cas.Expander == nil {
+		cas.Expander = fi.String("least-waste")
```
Comment (gjtempleton) [inline pkg/model/components/clusterautoscaler.go:40]: Is there a reason for deviating from the CA and setting `least-waste` as the default expander rather than `random`?

File: k8s/crds/kops.k8s.io_clusters.yaml [hunk 20]
```diff
@@ -335,6 +335,28 @@ spec:
...
+                properties:
+                  balanceSimilarNodeGroups:
+                    description: 'BalanceSimilarNodeGroups makes cluster autoscaler treat similar node groups as one. Default: false'
+                    type: boolean
+                  enabled:
+                    description: 'Enabled enables the cluster autoscaler. Default: false'
+                    type: boolean
+                  expander:
+                    description: 'Expander determines the strategy for which instance group gets expanded. Supported values: least-waste, most-pods, random. Default: least-waste'
```
Comment (gjtempleton) [inline k8s/crds/kops.k8s.io_clusters.yaml:348]: Is there a reason for not supporting `priority` and `price` expanders? 

Obviously would require some more docs/links out to CA docs in both cases.

File: pkg/model/components/clusterautoscaler.go (hunk 19 above)
Comment (olemarkus) [reply pkg/model/components/clusterautoscaler.go:40]: Think this was the default in the example I sourced from. No other specific reason.

File: k8s/crds/kops.k8s.io_clusters.yaml (hunk 20 above)
Comment (olemarkus) [reply k8s/crds/kops.k8s.io_clusters.yaml:348]: Only because of simplicity. Shouldn't be a problem to add this later if anyone needs this. I'd rather not have that block this PR at least

Comment (gjtempleton) [reply k8s/crds/kops.k8s.io_clusters.yaml:348]: That seems reasonable enough.

File: pkg/apis/kops/validation/validation.go [hunk 21]
```diff
@@ -1113,3 +1117,17 @@ func validateNodeLocalDNS(spec *kops.ClusterSpec, fldpath *field.Path) field.Err
 
 	return allErrs
 }
+
+func validateClusterAutoscaler(cluster *kops.Cluster, spec *kops.ClusterAutoscalerConfig, fldPath *field.Path) (allErrs field.ErrorList) {
+	if !cluster.IsKubernetesGTE("1.15") || cluster.IsKubernetesGTE("1.20") {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "Cluster autoscaler requires kubernetesVersion between 1.11 and 1.19"))
```
Comment (johngmyers) [inline pkg/apis/kops/validation/validation.go:1123]: I don't think there's any other validation that puts an upper bound on the Kubernetes version for a reason other than removal of support. Cilium, for example, allows the highest known Cilium version be used for arbitrary newer Kubernetes versions. So I don't think we should cap this either.

File: pkg/apis/kops/validation/validation.go [hunk 22]
```diff
@@ -1113,3 +1117,17 @@ func validateNodeLocalDNS(spec *kops.ClusterSpec, fldpath *field.Path) field.Err
...
+	}
+
+	allErrs = append(allErrs, IsValidValue(fldPath.Child("expander"), spec.Expander, []string{"least-waste", "random", "most-pods"})...)
+
+	if kops.CloudProviderID(cluster.Spec.CloudProvider) == kops.CloudProviderOpenstack {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "Cluster autoscaler is not supported on OpenStack"))
```
Comment (johngmyers) [inline pkg/apis/kops/validation/validation.go:1129]: I suspect it might not be supported on Spotinst either.

File: pkg/model/components/clusterautoscaler.go (hunk 19 above)
Comment (johngmyers) [reply pkg/model/components/clusterautoscaler.go:40]: We should probably take the default from cluster-autoscaler itself. That appears to be `random`.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 23]
```diff
@@ -526,3 +528,34 @@ func (tf *TemplateFunctions) OpenStackCCMTag() string {
 	}
 	return tag
 }
+
+// GetNodeInstanceGroups returns a map containing the defined instance groups.
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:532]:
```suggestion
// GetNodeInstanceGroups returns a map containing the defined instance groups of role "Node".
```

File: upup/pkg/fi/cloudup/template_functions.go [hunk 24]
```diff
@@ -526,3 +528,34 @@ func (tf *TemplateFunctions) OpenStackCCMTag() string {
...
+func (tf *TemplateFunctions) GetNodeInstanceGroups() map[string]kops.InstanceGroupSpec {
+	nodegroups := make(map[string]kops.InstanceGroupSpec)
+	for _, ig := range tf.KopsModelContext.InstanceGroups {
+		if ig.Spec.Role == kops.InstanceGroupRoleNode {
+			nodegroups[ig.ObjectMeta.Name] = ig.Spec
+		}
+	}
+	return nodegroups
+}
+
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:542]:
```suggestion

// ClusterAutoscalerTag returns the image tag for cluster-autoscaler.
```

File: upup/pkg/fi/cloudup/template_functions.go (hunk 3 above)
Comment (johngmyers) [reply upup/pkg/fi/cloudup/template_functions.go:557]: Please respond to this review comment.

File: pkg/apis/kops/validation/validation.go (hunk 21 above)
Comment (olemarkus) [reply pkg/apis/kops/validation/validation.go:1123]: The reason for this one was "remembering to add support for newer autoscaler". Original suggestion was to panic in the template function. Third option is to use the latest tag when we have an unknown/unsupported K8s version.

I suspect we'll bump the upper bound long before the first kops alpha that supports the new k8s version.

File: upup/pkg/fi/cloudup/template_functions.go (hunk 3 above)
Comment (olemarkus) [reply upup/pkg/fi/cloudup/template_functions.go:557]: So this is why I had the validation logic instead. I just felt it nicer should we ever forget to add support here.

File: pkg/apis/kops/validation/validation.go (hunk 22 above)
Comment (olemarkus) [reply pkg/apis/kops/validation/validation.go:1129]: I suspect the same. They still roll their own, it seems

Comment (olemarkus) [reply pkg/apis/kops/validation/validation.go:1129]: Doesn't seem like we treat spotinst fully as a cloudprovider. And cluster autoscaler will work for those IGs not managed by Spotinst. So I think it is best to leave it as it is for now. Spotinst is also behind a feature flag, so users know to take care.

Comment (johngmyers) [reply pkg/apis/kops/validation/validation.go:1129]: Fair enough

File: pkg/apis/kops/validation/validation.go (hunk 21 above)
Comment (hakman) [reply pkg/apis/kops/validation/validation.go:1123]: Could 1.13 and 1.14 also be supported, this would cover all supported versions? Also validation lower limit is for 1.15 and message says 1.11.
I also think that shouldn't be an upper limit and just use the one for the previous version.

Comment (olemarkus) [reply pkg/apis/kops/validation/validation.go:1123]: I just added the versions that had a release the last 6 months or so. Can add more, but I am not 100% sure the manifests will work that far back. 

My initial PR used latest tag for higher versions. But using latest release works for me too. Probably doesn't break anything.

Comment (hakman) [reply pkg/apis/kops/validation/validation.go:1123]: I am ok with latest as the default tag, as long as we add the rest of the images. Probably the tag should be changed to full image name to allow older image names.
The Helm chart seems to work for Kubernetes 1.8+ with the same manifest, so should be ok.

File: pkg/model/components/clusterautoscaler.go [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+			case 17:
+				tag = "v1.17.3"
+			case 16:
+				tag = "v1.16.6"
+			case 15:
+				tag = "v1.15.7"
+			}
+		}
+
+		cas.Image = fi.String("k8s.gcr.io/autoscaling/cluster-autoscaler:" + tag)
```
Comment (hakman) [inline pkg/model/components/clusterautoscaler.go:58]:
```suggestion
		var image string
		v, err := util.ParseKubernetesVersion(clusterSpec.KubernetesVersion)
		if err == nil {
			switch v.Minor {
			case 19:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.19.0"
			case 18:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.18.2"
			case 17:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.17.3"
			case 16:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.16.6"
			case 15:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.15.7"
			case 14:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:v1.14.8"
			case 13:
				image = "gcr.io/google-containers/cluster-autoscaler:v1.13.9"
			case 12:
				image = "gcr.io/google-containers/cluster-autoscaler:v1.12.8"
			default:
				image = "k8s.gcr.io/autoscaling/cluster-autoscaler:latest"
			}

			cas.Image = fi.String(image)
```

File: pkg/apis/kops/validation/validation.go (hunk 21 above)
Comment (hakman) [reply pkg/apis/kops/validation/validation.go:1123]: Now that you added the image override, I don't mind if we have the upper limit for this as long as it is ignored when the image is overridden. Your call.

**Comment by hakman** [comment]
@johngmyers leaving the approval to you, in case there is anything left from your point of view.
/lgtm

File: pkg/model/components/clusterautoscaler.go [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+func (b *ClusterAutoscalerOptionsBuilder) BuildOptions(o interface{}) error {
+	clusterSpec := o.(*kops.ClusterSpec)
+	cas := clusterSpec.ClusterAutoscaler
+	if cas == nil || !fi.BoolValue(cas.Enabled) {
+		return nil
+	}
+
+	if cas.Image == nil {
+
+		image := "k8s.gcr.io/autoscaling/cluster-autoscaler:latest"
```
Comment (johngmyers) [inline pkg/model/components/clusterautoscaler.go:42]: I'm not a big fan of mutable tags such as "latest"; I'd prefer using "v1.19.0" (or whatever the highest known version is) for newer Kubernetes versions.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 27]
```diff
@@ -521,6 +521,27 @@ func (b *BootstrapChannelBuilder) buildAddons() *channelsapi.Addons {
 		}
 	}
 
+	if b.Cluster.Spec.ClusterAutoscaler != nil && fi.BoolValue(b.Cluster.Spec.ClusterAutoscaler.Enabled) {
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:524]: Perhaps we should default `Enabled` to true if `ClusterAutoscaler` is non-`nil`

**Review by johngmyers — COMMENTED** [review]
Neither of the review comments is blocking. Holding to give opportunity to address.

/approve
/hold

File: pkg/model/components/clusterautoscaler.go (hunk 26 above)
Comment (hakman) [reply pkg/model/components/clusterautoscaler.go:42]: I don't have any preference here.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go (hunk 27 above)
Comment (hakman) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:524]: We agreed to keep it disabled by default at last office hours.
Maybe in 1.20 we materialise it into to the cluster spec so that people know it's enabled.

Comment (johngmyers) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:524]: But default is to have a `nil` `ClusterAutoscaler` field. Why would someone specify a `ClusterAutoscaler` field yet want `Enabled` to default to `false`?

**Comment by olemarkus** [comment]
I am a bit back and forth on what I prefer on the above. So I think I'll just let this one go in and then do a follow-up. Especially if cluster-autoscaler should prove to be slower with their support for newer k8s in the future.
/hold cancel

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go (hunk 27 above)
Comment (hakman) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:524]: I agree that it may be a good idea in general, just don't know how used this pattern is in kops.
Sometimes when I test or just want to disable something I just comment the `enabled: true` line, without actually but keep all other setting to make it easy to reenable.

One of the reasons for starting the discussion on https://github.com/kubernetes/kops/issues/9661 was to establish some guidelines for enabled/disabled and managed/unmanaged, but seems not very popular topic. Maybe should open a more generic one.

**Comment by paalkr** [comment]
Thanks for contributing with this addition, that makes deploying the CAS easier. We have a use case that uses a mix of different cluster AutoScaling methods (external and CAS) for various node pools. I think it would be vice to add a keyword to the instance group spec to either include or exclude the group with CAS.

**Comment by olemarkus** [comment]
Thanks. I'd add this as a separate issue. I may not be able to implement this in the near future, but it should be a pretty easy "good first issue".

**Comment by avdhoot** [comment]
Might late to join the party. Thanks @olemarkus for this feature. It will remove extra step to manage  `autoscaler` version separately .  But it will be good if we add support for `priority`  expander.  When you have multiple instance group mix of on-demand & spot instance group. `priority`  expander help to set priority to spot instance group.
